File: src/test/java/logic/FindTest.java
package logic;

import java.io.File;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class FindTest {

    private Throwable exception;

    @Test
    void test1() throws Exception {
        Find path = new Find("src", true);
        assertEquals(new File("src/test/test1.txt"),
                path.find("test1.txt"));
    }

    @Test
    void test2() throws Exception {
        Find path = new Find("src/test", false);
        assertEquals(new File("src/test/test1.txt"),
                path.find("test1.txt"));
    }

    @Test
    void test3() {
        Find path = new Find("src", false);
        exception = assertThrows(Exception.class, () -> path.find("test1.txt"));
        assertEquals("File does not exist", exception.getMessage());
    }

    @Test
    void test4() throws Exception {
        Find path = new Find("src", false);
        assertEquals(new File("src/test2.txt"),
                path.find("test2.txt"));
    }

    @Test
    void test5() {
        Find path = new Find("src/test", true);
        exception = assertThrows(Exception.class, () -> path.find("test3.txt"));
        assertEquals("File does not exist", exception.getMessage());
    }

}

File: src/test/kotlin/logic/FindTest.kt
package logic

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.nio.file.Path
import java.nio.file.Paths

class FindTest {
    val testdir = Paths.get("").toRealPath().resolve("src/test/testdir")

    @Test
    fun find() {
        generateTestFiles()

        assertEquals(null, Find().find(false, null, "code.kt"))

        //  -r
        assertEquals(setOf(
                testdir.resolve("anotherProgramming/kotlin/code.kt").toFile(),
                testdir.resolve("programming/kotlin/code.kt").toFile(),
                testdir.resolve("programming/kotlin/deeper/code.kt").toFile()),
                Find().find(true, null, "code.kt")?.toSet())

        //  -d
        assertEquals(setOf(testdir.resolve("programming/kotlin/code.kt").toFile()),
                Find().find(false, testdir.resolve("programming/kotlin").toFile(), "code.kt")?.toSet())

        //  -r -d
        assertEquals(setOf(
                testdir.resolve("programming/kotlin/code.kt").toFile(),
                testdir.resolve("programming/kotlin/deeper/code.kt").toFile()),
                Find().find(true, testdir.resolve("programming").toFile(), "code.kt")?.toSet())

        //  find directories
        assertEquals(setOf(
                testdir.resolve("anotherProgramming/kotlin").toFile(),
                testdir.resolve("programming/kotlin").toFile()),
                Find().find(true, testdir.toFile(), "kotlin")?.toSet())

        deleteTestFiles()
    }

    fun generateTestFiles() {
        createDirectory(testdir)
        createDirectory(testdir.resolve("anotherProgramming"))
        createDirectory(testdir.resolve("anotherProgramming/kotlin"))
        testdir.resolve("anotherProgramming/kotlin/code.kt").toFile().createNewFile()
        createDirectory(testdir.resolve("programming"))
        createDirectory(testdir.resolve("programming/java"))
        testdir.resolve("programming/java/code.java").toFile().createNewFile()
        createDirectory(testdir.resolve("programming/kotlin"))
        testdir.resolve("programming/kotlin/code.kt").toFile().createNewFile()
        createDirectory(testdir.resolve("programming/kotlin/deeper"))
        testdir.resolve("programming/kotlin/deeper/code.kt").toFile().createNewFile()
    }

    fun deleteTestFiles() {
        testdir.toFile().deleteRecursively()
    }

    fun createDirectory(path: Path) {
        val file = path.toFile()
        if (!file.exists() || !file.isDirectory) {
            file.mkdir()
        }
    }
}